tests/test_optimizer/test_dist_lamb.py

```diff
@@ -0,0 +1,319 @@
+    """
+    rank = dist.get_rank(group)
+    world_size = dist.get_world_size(group)
+
+    for p, torch_p in zip(dist_module.parameters(), torch_model.parameters()):
+        if torch_p.grad is None:
+            # avoid inconsistent grad and param dtype error
+            force_assign_grad(torch_p, g_dtype)
+        else:
+            torch_p.grad += torch.randn_like(torch_p, device=torch_p.device, dtype=g_dtype)
+
+        if p.grad is None:
+            force_assign_grad(p, g_dtype)
+
+        if is_distributed_tensor(p):
+            split_dim = get_split_dim(p)
+            # Add grads only to the correctly split chunk
+            force_assign_grad(p, g_dtype, torch_p.grad.chunk(world_size, dim=split_dim)[rank])
+            # assert_close(p.grad, torch_p.grad.chunk(world_size, dim=split_dim)[rank])
+        else:
+            force_assign_grad(p, g_dtype, torch_p.grad)
+
+
+@parameterize("p_g_dtype", _ALLOWED_P_G_TYPES)
+@parameterize("bias_correction", [False, True])
+@parameterize("tp_zero_size", [(1, 4), (4, 1), (2, 2)])
+def run_dist_lamb_basic(
+    bias_correction: bool, p_g_dtype: tuple[torch.dtype, torch.dtype], tp_zero_size: tuple[int, int]
+) -> None:
+    """Test without forward"""
+    p_dtype, g_dtype = p_g_dtype
+    tp_size, zero_size = tp_zero_size
+
+    # Set distributed groups
+    rank = dist.get_rank()
+    clear_layout_converter()  # Ensure correct sharding
+    proc_mesh = ProcessGroupMesh(tp_size, zero_size)
+    tp_group = proc_mesh.get_group_along_axis(0)
+
+    tp_rank = dist.get_rank(tp_group)
+    torch.cuda.manual_seed(_SEED)  # Fix model init
+    torch_model = Net(in_dim=_IN_DIM, hid_dim=_HID_DIM, identity=True).to(rank)
+    tp_model = TPNet(torch_model.fc0, torch_model.fc1, torch_model.fc2, tp_group).to(rank)
+    # Ensure equal weight init
+    assert_close(
+        torch_model.fc1.weight[tp_rank * _HID_DIM // tp_size : (tp_rank + 1) * _HID_DIM // tp_size],
+        tp_model.fc1.weight,
+    )
+    assert_close(
+        torch_model.fc2.weight[:, tp_rank * _HID_DIM // tp_size : (tp_rank + 1) * _HID_DIM // tp_size],
+        tp_model.fc2.weight,
+    )
+
+    # Set up optimizers
+    lr = 1e-3
+    beta1, beta2 = 0.9, 0.999
+    eps = 1e-8
+    torch_optim = Lamb(
+        setup_param_groups(torch_model), lr=lr, betas=(beta1, beta2), eps=eps, bias_correction=bias_correction
+    )
+    optim = DistributedLamb(
+        setup_param_groups(tp_model),
+        lr=lr,
+        betas=(beta1, beta2),
+        eps=eps,
+        bias_correction=bias_correction,
+    )
+    optim.setup_distributed(tp_group)
+
+    rtol, atol = 8e-5, 8e-5
+    if p_dtype is torch.float16 or g_dtype is torch.float16:
+        rtol, atol = 2e-4, 2e-4
+    if p_dtype is torch.bfloat16 or g_dtype is torch.bfloat16:
+        rtol, atol = 4e-4, 4e-4
+
+    for i in range(_N_STEP):
+        torch.cuda.manual_seed(_SEED)  # NOTE: having only one manual_seed above doesn't work?
```

Comment:
Use this function:
https://github.com/hpcaitech/ColossalAI/blob/079bf3cb26a502fc647b1aad15fd14d6266be66c/colossalai/testing/random.py#L7

And for each step, the random seed should be different. You may use `_SEED + i`

Reply:
fixed